File: tiny-url/tiny-url.py
class TinyUrl:
    def __init__(self):
        self.long_short = {}
        self.short_long = {}
        self.char_dict = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        self.tot = 0

    # @param {string} url a long url
    # @return {string} a short url starts with http://tiny.url/
    def longToShort(self, url):
        # Write your code here
        if url not in self.long_short:
            tiny = ""
            tmp = self.tot
            for i in xrange(6):
                tiny += str(self.char_dict[tmp % 62])
                tmp /= 62
            self.tot += 1
            self.long_short[url] = tiny
            self.short_long[tiny] = url
        return "http://tiny.url/" + self.long_short[url]

    # @param {string} url a short url starts with http://tiny.url/
    # @return {string} a long url
    def shortToLong(self, url):
        # Write your code here
        if url[-6:] in self.short_long:
            return self.short_long[url[-6:]]
        else:
            return None

File: tiny-url/tiny-url.cc
class TinyUrl {
public:
    /**
     * @param url a long url
     * @return a short url starts with http://tiny.url/
     */
    string longToShort(string& url) {
        // Write your code here
        if (long_short.find(url) == long_short.end()) {
            string tiny = "000000";
            int tmp = tot;
            for (int i = 0; i < 6; ++i) {
                tiny[i] = dic[tmp % 62];
                tmp /= 62;
            }
            long_short[url] = tiny;
            short_long[tiny] = url;
            ++tot;
        }
        return domain + long_short[url];
    }
    
    /**
     * @param url a short url starts with http://tiny.url/
     * @return a long url
     */
    string shortToLong(string& url) {
        // Write your code here
        string sub = url.substr(url.length() - 6);
        if (short_long.find(sub) != short_long.end())
            return short_long[sub];
        else
            return NULL;
    }
    
private:
    const string domain = "http://tiny.url/";
    static int tot;
    const string dic = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    map<string, string> long_short;
    map<string, string> short_long;
};

int TinyUrl::tot = 0;

